setup.ps1

```diff
@@ -0,0 +1,229 @@
+if ($MyInvocation.InvocationName -ne $MyInvocation.MyCommand.Name) {
+  exit
+}
+
+# Check for admin privileges and elevate if necessary
+if (-NOT ([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]::Administrator)) {
+  Write-Log "Requesting administrative privileges..." -Level 'ERROR'
+  Start-Process PowerShell -ArgumentList "-File `"$PSCommandPath`"" -Verb RunAs
+  Exit-Script -exitCode 1 -isSubShell $false
+}
+
+# Log file setup
+"Admin rights confirmed" | Out-File $LogFilePath -Append
+"Starting the script operations..." | Out-File $LogFilePath -Append
+
+# Navigate to the project directory
+Set-Location -Path $ProjectDir
+"Current directory: $(Get-Location)" | Out-File $LogFilePath -Append
+
+# Define the path to the Python executable in the virtual environment
+$VenvPython = Join-Path $ProjectDir "$VenvName\Scripts\python.exe"
+
+# Check if the virtual environment exists, if not, create it
+if (-Not (Test-Path $VenvPython)) {
+  Write-Log "Virtual environment not found. Creating virtual environment..." -IsError $false
+  python -m venv "$VenvName"
+  if (-Not (Test-Path $VenvPython)) {
+    Write-Log "Failed to create virtual environment." -Level 'ERROR'
+    Exit-Script -exitCode 1
+  }
+  Write-Log "Virtual environment created successfully." -IsError $false
+}
+
+# Define the pip command using the Python executable
+$VenvPip = @($VenvPython, '-m', 'pip')
+
+# Activate the virtual environment
+$OldVirtualPath = $env:PATH
+$env:PATH = "$ProjectDir\$VenvName\Scripts;$env:PATH"
+
+# Ensure setuptools is installed using the virtual environment's pip
+Write-Log "Ensuring setuptools is installed..."
+& $VenvPip[0] $VenvPip[1..$VenvPip.Length] 'install', '-v', 'setuptools' | Out-File $LogFilePath -Append 2>&1
+if ($LASTEXITCODE -ne 0) {
+  Write-Log "Failed to install setuptools." -Level 'ERROR'
+  Exit-Script -exitCode 1
+}
+
+# Pull latest updates
+Write-Log "Pulling latest updates..."
+& git pull | Out-File $LogFilePath -Append 2>&1
+if ($LASTEXITCODE -ne 0) {
+  Write-Log "Failed to pull updates from Git." -Level 'ERROR'
+  Exit-Script -exitCode 1
+}
+
+# Install TA-Lib using the virtual environment's pip
+Write-Log "Installing TA-Lib using virtual environment's pip..."
+& $VenvPip[0] $VenvPip[1..$VenvPip.Length] 'install', '--find-links=build_helpers\', '--prefer-binary', 'TA-Lib' | Out-File $LogFilePath -Append 2>&1
+
+# Present options for requirement files
+$selectedIndices = Get-UserSelection -prompt "Select which requirement files to install:" -options $RequirementFiles -defaultChoice 'A'
+
+# Install selected requirement files
+foreach ($index in $selectedIndices) {
+  if ($index -lt 0 -or $index -ge $RequirementFiles.Length) {
+    Write-Log "Invalid selection index: $index" -Level 'ERROR'
+    continue
+  }
+  
+  $filePath = Join-Path $ProjectDir $RequirementFiles[$index]
+  if (Test-Path $filePath) {
+    Install $filePath
```

Comment:
Do i understand this section right that if i insert "A,B,C" - then it'll install all 3 requiements files independently?

You should rather cache the results - and then run `-r requirements.txt -r requirements-dev.txt`, ... 
that'll make the installation process a lot faster.

Reply:
Is this internally optimizing the way it downloads the requirements?

Reply:
if you separately install `requirements.txt` and `requirements-dev.txt` - then all dependencies from `requirements.txt` will need to be checked / verified if the version fits, if it's already installed, ...) twice.
that's because most requirements files are wrapped within each other (requirements-dev.txt references all other files, ...).

It's not a download topic as much - as pip will be clever enough to see that the dependency is already installed - but there's still work needed to check for the proper versions, ...

In some cases, non-pinned (floating) dependencies can have different version constraints - so requirements.txt might install "ABC==2.4.1" (the latest version) - but a dependecy in `requirements-dev.txt` requires `ABC<=2.4.0` - so ABC is removed, and reinstalled at 2.3.4 (or whatever version).